src/pybind/mgr/volumes/subvolume.py

```diff
@@ -0,0 +1,275 @@
+        self.fs.rmdir(path)
+
+    def list_groups(self):
+        """
+        Returns a list of subvolume groups within a FS volume/filesystem
+        """
+        spec = SubvolumeSpec("", "")
+        path = spec.group_dir
+        return self._list_dir(path)
+
+    def get_group_path(self, spec, create_default=False):
+        path = spec.group_path
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            if create_default and spec.is_default_group():
+                self._mkdir_p(path)
+                return path
+            return None
+        return path
+
+    def _get_ancestor_xattr(self, path, attr):
+        """
+        Helper for reading layout information: if this xattr is missing
+        on the requested path, keep checking parents until we find it.
+        """
+        try:
+            result = self.fs.getxattr(path, attr).decode('utf-8')
+            if result == "":
+                # Annoying!  cephfs gives us empty instead of an error when attr not found
+                raise cephfs.NoData()
+            else:
+                return result
+        except cephfs.NoData:
+            if path == "/":
+                raise
+            else:
+                return self._get_ancestor_xattr(os.path.split(path)[0], attr)
+
+    def _snapshot_create(self, snappath, mode=0o755):
+        """
+        Create a snapshot, or do nothing if it already exists.
+        """
+        try:
+            self.fs.stat(snappath)
+        except cephfs.ObjectNotFound:
+            self.fs.mkdir(snappath, mode)
+        else:
+            log.warn("Snapshot '{0}' already exists".format(snappath))
+
+    def _snapshot_delete(self, snappath):
+        """
+        Remove a snapshot, or do nothing if it doesn't exist.
+        """
+        self.fs.stat(snappath)
+        self.fs.rmdir(snappath)
+
+    def create_subvolume_snapshot(self, spec, snapname, mode=0o755):
+        snappath = spec.make_subvol_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_create(snappath, mode)
+
+    def remove_subvolume_snapshot(self, spec, snapname):
+        snappath = spec.make_subvol_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_delete(snappath)
+
+    def create_group_snapshot(self, spec, snapname, mode=0o755):
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_create(snappath, mode)
+
+    def remove_group_snapshot(self, spec, snapname):
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        return self._snapshot_delete(snappath)
+
+    def connect(self):
+        log.debug("Connecting to cephfs...")
+        self.fs = cephfs.LibCephFS(rados_inst=self.rados)
+        log.debug("CephFS initializing...")
+        self.fs.init()
+        log.debug("CephFS mounting...")
+        self.fs.mount(filesystem_name=self.fs_name)
```

Comment:
See, https://github.com/ceph/ceph/pull/28380/commits/a203e577098d82a6d0dea86757c9e83dccdba78d

Reply:
already did that when rebasing yesterday (and another similar change in `setxattr`)